### DISCUSSION TITLE
Add localised wordlists to allow passpharase generation in users' native languages.

### DISCUSSION THREAD

**Discussion by FabianVolkers** [bitwarden/clients#12014] [Password Manager]
### Code Contribution Proposal

**As a user I want to generate a passphrase in my native language, So I can easily remember and type it.**

This user story results out of user feedback at a company in a non English-speaking country. For situations where passwords cannot be copy and pasted from the Bitwarden client but instead have to be transferred manually to a different system, users require words they know.

My proposal is adding localised word lists and adding language to the passphrase generation options. Code changes for the generation seem to be limited to `libs/tools/generator` and `libs/common/src/platform`. By keeping English as the default, the localised generation feature is introduced as a non-breaking change. In the client, the option should be presented as a dropdown of available languages.

Considering the dice logic used for passphrase generation, localised word lists need to be selected based on the right criteria. For the time being, I would implement the German word list from https://github.com/ulif/diceware

Are there any caveats I need to consider before implementing this change?

Related discussion and feature requests:
- https://community.bitwarden.com/t/choose-the-word-list-for-passphrases/3372/54
- https://github.com/orgs/bitwarden/discussions/5512

**Comment by HydraheadHunter** [comment]
I'd like to second this suggestion, though I would ammend the suggestion slightly. 

Rather than have the dropdown menu that lets users choose between English, German, Spanish, etc. it should let users choose which language(s) to include in the word-pool: for example, ✓English, ✓German, ✗Spanish would generate phrases that are a blend of english and german words.

The default setting to introduce this as a non breaking change would then be ✓English, ✗German, ✗Spanish, ✗etc rather than English.

And yes to the custom words: I'm from the site of fine shoelaces and there some words in my lexicon that ain't make it into nobody's lib and dicts.

**Comment by bwbug** [comment]
> For the time being, I would implement the German word list from https://github.com/ulif/diceware
>Are there any caveats I need to consider before implementing this change?

I'm just a Bitwarden customer, but I would recommend allowing users to use their own word lists.  If you are going to hardcode a word list into the Bitwarden clients, then I would presume that Bitwarden would want to vet the word list to ensure it contains no offensive words.   The word list you have proposed implementing ([wordlist_de.txt](https://github.com/ulif/diceware/blob/master/diceware/wordlists/wordlist_de.txt) from [github.com/ulif/diceware](https://github.com/ulif/diceware/tree/master/diceware/wordlists)) contained the n-word until [2022](https://github.com/ulif/diceware/commit/768466886d0b67702fd563bdd42564543c22cd36), and when removing that word, the developer conceded that "there are certainly more words to remove and we must face the problem, that it is not easy to generate proper welcoming wordlists automatically".

By allowing users to supply their own word lists, you step around this problem.

**Comment by Sythelux** [comment]
also want to opt in on this. I installed bitwarden for my mom, who doesn't speak English very well and would really like bitwarden to use words of her native language. I don't actually really care about the exact solution, but the passphrase isn't really helpful atm.

**Comment by bwbug** [comment]
@Sythelux In the meantime, your mom (or anybody else who wants non-English passphrases) could use one of the first three tools ("Alternate", "Cryptocurrency", or "Diceware") on this webpage:

https://atoponce.github.io/webpassgen/

Any of those three have non-English options for the language selector.

**Comment by dede34fr** [comment]
A data point from a self-hosted (Vaultwarden) family deployment in a French-speaking household, plus one measurement that may save the next person some time.

**Where the English list actually hurts.** Not inside the vault — autofill makes the wordlist irrelevant there. It hurts for the credentials that get *typed by hand or read aloud*: a smart-TV sign-in, a game console, a Wi-Fi guest key dictated over the phone to an elderly relative. Those are exactly the cases where a passphrase beats a random string, and exactly where an English list fails: `abacus-zoologist-abridge` is neither memorable nor reliably typeable for a non-English speaker, so people fall back to a weak password they made up themselves. The security loss is real, and it happens outside the client where no telemetry will ever show it.

**The entropy question is a non-issue.** The EFF long list is 7,776 words = 12.925 bits/word. Any curated list of the same size, in any language, yields exactly the same entropy. This is localisation, not a security trade-off — and with English staying the default, it is non-breaking by construction.

**The real work is licensing and curation, not code.** Worth stating explicitly, because the two French lists most often suggested as "just use this one" both fail on inspection (checked today):

- `chmduquesne/diceware-fr` — licence is fine (CC-BY-3.0), but the 7,776 entries are not all words: entry `11111` is `=`, `11112` is `==`, `66666` is `zzzz`. Unusable as shipped.
- `mbelivo/diceware-wordlists-fr` — 7,776 genuine words (`abaisse` → `zoom`), but there is no LICENSE file, and the README states the source lexicon (Lexique) is **CC BY-NC-SA 3.0**, with the author explicitly unsure whether that propagates to a derived list. The NC clause alone rules it out for Bitwarden.

So "adopt an existing list" is not actually available, at least for French. Whichever team picks this up will need either a licence-clean generation pass over a permissively licensed corpus, or a per-language curation effort. That is a bigger ask than this thread has assumed, and it probably explains why the forum request has been open since December 2018 — which is an argument for scoping it properly, not for leaving it open another eight years.

One note for self-hosters, since this comes up: the list is hard-coded in the clients (checked in the shipped web vault bundle — 7,776 entries, `abacus` → `zoologist`), and the CLI has no language option either. There is no local workaround; it can only be fixed upstream.

**Reply by bwbug** [reply]
@dede34fr  What about the [French Wordlist by Tango for Tails/TOR](https://theworld.com/~reinhold/wordlist_fr_7776.txt), which is available under a CC0 1.0 Universal license ([README](https://theworld.com/~reinhold/Tails_fr_lists_README.txt))?
